modules/optflow/src/block_matching.cpp

```diff
@@ -0,0 +1,143 @@
+ // are permitted provided that the following conditions are met:
+ //
+ //   * Redistribution's of source code must retain the above copyright notice,
+ //     this list of conditions and the following disclaimer.
+ //
+ //   * Redistribution's in binary form must reproduce the above copyright notice,
+ //     this list of conditions and the following disclaimer in the documentation
+ //     and/or other materials provided with the distribution.
+ //
+ //   * The name of the copyright holders may not be used to endorse or promote products
+ //     derived from this software without specific prior written permission.
+ //
+ // This software is provided by the copyright holders and contributors "as is" and
+ // any express or implied warranties, including, but not limited to, the implied
+ // warranties of merchantability and fitness for a particular purpose are disclaimed.
+ // In no event shall the Intel Corporation or contributors be liable for any direct,
+ // indirect, incidental, special, exemplary, or consequential damages
+ // (including, but not limited to, procurement of substitute goods or services;
+ // loss of use, data, or profits; or business interruption) however caused
+ // and on any theory of liability, whether in contract, strict liability,
+ // or tort (including negligence or otherwise) arising in any way out of
+ // the use of this software, even if advised of the possibility of such damage.
+ //
+ //M*/
+
+#include "precomp.hpp"
+
+namespace cv
+{
+namespace optflow
+{
+
+class OpticalFlowBlockMatching : public DenseOpticalFlow
+{
+protected:
+  int windowSize;
+  int blockSize;
+
+  inline float submatrixAbsDiff( int x0, int y0, const Mat &I0, int x1, int y1, const Mat &I1 ) const;
+
+public:
+  OpticalFlowBlockMatching() : windowSize( 3 ), blockSize( 8 ){};
+
+  void calc( InputArray I0, InputArray I1, InputOutputArray flow );
+  void collectGarbage();
+};
+
+inline float OpticalFlowBlockMatching::submatrixAbsDiff( int x0, int y0, const Mat &I0, int x1, int y1,
+                                                         const Mat &I1 ) const
+{
+  float error = 0;
+  const Size size = I0.size();
+  for ( int i = -windowSize; i <= windowSize; ++i )
+  {
+    if ( i + y0 < 0 || i + y0 >= size.height || i + y1 < 0 || i + y1 >= size.height )
+    {
+      error += 1;
+      continue;
+    }
+    const Vec3f *I0X = I0.ptr<Vec3f>( i + y0 );
+    const Vec3f *I1X = I1.ptr<Vec3f>( i + y1 );
+    for ( int j = -windowSize; j <= windowSize; ++j )
+    {
+      if ( j + x0 < 0 || j + x0 >= size.width || j + x1 < 0 || j + x1 >= size.width )
+      {
+        error += 1;
+        continue;
+      }
+      const Vec3f diff = I0X[j + x0] - I1X[j + x1];
+      error += abs( diff[0] );
+      error += abs( diff[1] );
+      error += abs( diff[2] );
+    }
+  }
+  return error;
+}
+
+void OpticalFlowBlockMatching::calc( InputArray I0, InputArray I1, InputOutputArray flow_out )
+{
+  CV_Assert( I0.channels() == 3 );
```

Comment:
I think this "block matching" shouldn't be here at all :) I wrote it for a testing purpose to get acquainted with the existing optflow codebase. This is not supposed to be used by anyone else.